Allow to restore a metadata record from deleted records (Part 3)

ianwallen opened the pull request:
This pull request is a continuance of pull request https://github.com/geonetwork/core-geonetwork/pull/4751. It contains the changes from that pull request and I will merge the changes once pull request https://github.com/geonetwork/core-geonetwork/pull/4751 has been completed.  Marking this as WIP until https://github.com/geonetwork/core-geonetwork/pull/4751 has been completed.

**Changes in this pull request**
- Changed the primary key for the metadatastatus table so that it is a single key instead of a multikey.  Kept the multikey fields as a unique index. 
- With new PK added, a new relatedmetadatastatusid field was also added so that a restored item can point to the parent for which the restore was done from (see #4750 )
- To support being able to view the history listing of deleted items (which no longer exists in index), the UUID and Titles were also added to the metadata status table.   The title field  is stored as a json representation of a hashmap containing the language and titles.
- To support getting the multilingual titles from the metadata, the metadata plugin now require a new extract-titles.xsl file to get the tilles.  If it does not exists, it will only use the index values so if a record is deleted and the index entry does not exists then it will not be possible to get the title so it will be empty in the status listing.
- Added groupownname to geonet:info to assist with the recovery as we needed to be able to restore the record to the same group that it was deleted from.
- Updated API parameters and results to support new table changes.

**Database Migration**
The expected database upgrade process should look something like the following

- Start geonetwork
- JPA will attempt to apply all the database changes and there will be errors.  The errors are caused because JPA cannot figure out how to change the primary keys. So it simply attempts to add a new PK for ID but there is already the old PK.  Also since it cannot create the PK. it also fails attempting to create a foreign key for the relatedmetadatastatusid.
- The migration java class(UpdateMetadataStatus) will attempt to update the primary key by adding fields as null, update data, and then make it not null and then drop/recreate the pk.  The drop recreate PK may generate some errors as well as it attempts to drop the constraint, index pk (hopefully this will cover all database vendors). I in the end the pk should be removed. and re-created.
- After the system has started, it is recommended that the system be shut down and restarted. This will fix the missing foreign keys and possibly other missing changes that failed on first startup.  It should not be a required task as the system should still be able to work without the foreign key.
- On the next system restart, JPA should finish any missing changes.  We should not see database errors any errors on the second restart. Also the tables should have all been changed after this point and the migration should be complete.  

**Todo**
-  [x] -  Change some status API's to support new single metadata status ID key.
-  [x] -  Migration - primary key issue  - did not figure out how to remove old primary key from table - current attempt fails on h2 database.

**Whats next**

 -  Updated user interface so it is possible to see the linkage from a restored item using the new relatedmetadatastatus.
 -  Improve restore so that other items, (such as categories, rating, permissions, resources...),  are included in the restore.
 -  Migration - title issue - see question below - I was not able to get the current titles. If not part of the migration then old data will contain null titles.  This is not really required but will provide a smoother upgrade path.  

**Dependencies**
https://github.com/geonetwork/core-geonetwork/pull/4751 (contains previous versioning work which is included in this PR but is a smaller subset of changes.  If this PR is approved then #4751 can be closed.)
https://github.com/geonetwork/core-geonetwork/pull/4781 (new key uses new sequence logic - so this is moreless required for consistency) 

**Questions**
For migration
   Should we make the titles mandatory? This would require looping thought all metadata status and fetching and assigning the current titles. This will be difficult to do in the future once some records have been deleted.
   Should we fail if the extra-titles.xls fail to parse or get the titles, or should we just allow null titles?  If required then all schema pluging will not be backwards compatible and will need extra-titles.xls
   I attempted to get the titles as part of the migration however it fails as the schema's don't seem to be loaded prior to migration. I also attempted to get the titles from the index however the index was also not available at the time of migration so I used getTitle from the metadata which is mostly null and I think it is depreciated. (see commented code in migration script for details https://github.com/geonetwork/core-geonetwork/pull/4817/files#diff-6453eb82272709dfbed89c0ba612c9f9R214-R260)

Commit:
Restore from history functionality

Commit:
Behavior when draft exists

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory

# Conflicts:
#	web-ui/src/main/resources/catalog/locales/en-core.json
#	web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v390/migrate-default.sql

ianwallen committed:
Fixed migration merge

ianwallen committed:
Removed MetadataStatusId Serialization and Deserialization and used existing format of /{metadataUuid}/status/{statusId:[0-9]+}.{userId:[0-9]+}.{changeDate}
Removed rollback option as a rollback is the same as a restore.  Also removed logic to delete status of rollback status.
Renamed view to view previous version and changed version and added new showStatusAfter api to support this.
Database
Added missing fre language for insert
Move migration inserts to 3110 and also added missing translation.

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory

# Conflicts:
#	services/src/main/java/org/fao/geonet/api/records/model/MetadataStatusResponse.java

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory

# Conflicts:
#	web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v3110/migrate-default.sql

ianwallen committed:
Allow capability to recover deleted records
Changed PK key for metadatastatus so that it is a single key.
Added groupownname to geonet:info to assist with the recovery.
Add metadataStatus entry for deleted objects.
    Added UUID and Titles to the status so that the data is available for deleted records as well.

ianwallen committed:
removed duplicate status_id which is the same as the statusvalue.

ianwallen committed:
Fixed improper resolved conflict which accidentally converted file to ascci causing script to fail.

ianwallen committed:
Updated migration script
attempted to load titles but failed.

jodygarnett commented:
@ianwallen this change (replacing primary) key is pretty big change to backport, consider target for 3.11.x (released as 3.12.x)

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory_pk_fix

ianwallen committed:
More versioning changes
Changed titles to LinkedHashMap so that the order is static.
Updated status search api so that we can query based on id and uuid.
Updated API results
   Changed RelatedMetadataStatus to only return the id instead of the object.
   Return titles as a json object (applied temp hack until we upgrade jpa)
   renamed *statusEmpty to *stateEmpty for consistency
   Fixed bug with some fields being null when they should have values.
   Removed *statusDetail as we already had the *state fields from the parent.  If we want them named statusDetail then we should rename the fields in MetadataStatus.
Also updated migration script to attempt to fix primary key issue - but may need more work.

ianwallen committed:
Added @Type annotations

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory_pk_fix

ianwallen committed:
Fixed H2 migration primary key issues and also completed the rest field migrations issues (except for the the titles)
    Removed multikey Unique key on metadata status.  It was confusing H2 as it was using the same index for the PK and unique key making it difficult to remove the primary key.
Added missing commands in the oracle and mysql scripts.

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory_pk_fix

# Conflicts:
#	web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v3110/migrate-default.sql

ianwallen commented:
@jodygarnett I agree that this may be big to backport so I'm find to target this to the next major release (3.12 it it will exists or 4.0 if there will not be a 3.12)

ianwallen commented:
Also since this PR contains all the changes from #4751.  If this PR is reviewed and accepted then #4751 can be cancelled.  But if #4751 is accepted first then we can merge any changes back to this one.  The main reason they were split was to make it possible to review smaller pieces.

jodygarnett commented:
@ianwallen the "Whats next" checkmarks make it unclear if this PR is ready to be applied. Given that you continue the work later I would like to assume you are breaking up this activity in a number of pull requests.

josegar74 commented on services/src/main/java/org/fao/geonet/api/records/MetadataWorkflowApi.java:
```diff
@@ -419,23 +442,268 @@ public MetadataStatus convertParameter(int id, MetadataStatusParameter parameter
+                            userId, changeDate));
+        }
+
+        String fullMetadataText = metadataStatus.getCurrentState();
```
To check to refactor with the code https://github.com/geonetwork/core-geonetwork/pull/4817/files#diff-668fdf0a09c1a8afb5771c9e85715598R496-R506, seem the same code

ianwallen replied:
refactored code for before/after api so that that they use common code.

josegar74 commented on services/src/main/java/org/fao/geonet/api/records/MetadataWorkflowApi.java:
```diff
@@ -419,23 +442,268 @@ public MetadataStatus convertParameter(int id, MetadataStatusParameter parameter
+                            userId, changeDate));
+        }
+
+        AbstractMetadata metadata = null;
```
To check to refactor with the code https://github.com/geonetwork/core-geonetwork/pull/4817/files#diff-668fdf0a09c1a8afb5771c9e85715598R496-R506, seem the same code

ianwallen replied:
refactored code for before/after api so that that they use common code.

josegar74 commented on web-ui/src/main/resources/catalog/components/history/GnHistoryService.js:
```diff
@@ -35,17 +35,22 @@
         this.delete = function (step) {
-          return $http.delete('../api/records/' + step.id.metadataId + '/status/' +
-            step.id.statusId + '.' + step.id.userId + '.' + step.id.changeDate.dateAndTime);
+          return $http.delete('../api/records/' + step.uuid + '/status/' +
```
Not sure how this works with the workflow enabled and the draft metadata copies, as there're 2 metadata with different id, but same uuid.

ianwallen replied:
I have reverted this back to use the id.

josegar74 reviewed (commented):
Some small changes to review, also will require to rebase with `master`.

Something that will require improvement, but can be done as part of other PR is the restoration of files attached to the metadata. For now seem only stored the xml. When deleted it's stored also a MEF version in the data directory, to consider in the future to restore from that MEF (or store it in the database).

ianwallen committed:
Revert grouprepository findOne as it was not required.
Fixed indentation
Get first language in case there are multiple for the dc:language.

ianwallen commented on schemas/dublin-core/src/main/plugin/dublin-core/extract-titles.xsl:
```diff
@@ -0,0 +1,39 @@
+  <xsl:template match="simpledc">
+    <titles>
+      <title>
+        <xsl:attribute name="lang"><xsl:value-of select="dc:language"/></xsl:attribute>
```
Fixed

ianwallen commented on listeners/src/main/java/org/fao/geonet/listener/metadata/draft/DraftUtilities.java:
```diff
@@ -111,12 +112,17 @@ public AbstractMetadata replaceMetadataWithDraft(AbstractMetadata md, AbstractMe
+            st.setMetadataId(md.getId());
+            st.setUuid(md.getUuid());
+            try {
+            st.setTitles(metadataUtils.extractTitles(Integer.toString(md.getId())));
```
fixed

ianwallen commented on domain/src/main/java/org/fao/geonet/repository/GroupRepositoryCustom.java:
```diff
@@ -36,6 +37,16 @@
+     * @return the group with the given groupid
+     */
+    @Nullable
+    Group findOne(@Nonnull String groupId);
```
Looks like I did not need it at all as it was starting with an numeric id.
Reverted the changes and fixed code to use numeric id.

ianwallen commented:
@jodygarnett 
> @ianwallen the "Whats next" checkmarks make it unclear if this PR is ready to be applied. Given that you continue the work later I would like to assume you are breaking up this activity in a number of pull requests.

These a features that will need to be worked on in a future PR as this one is big enough.  I have removed the checkboxes to make it clearer that it will not be done as part of this PR.

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory_pk_fix

ianwallen committed:
Another fix for groupowner - for sample data which does not belong to a group.
Renamed deleteAllByMetadataId back to deleteAllById_MetadataId - There are lots of others that should be renamed and this will be done as separate pr.
Revert api calls using uuid as it could cause issues with drafts.
    Updated api to support metadataUuid being a metadata id or uuid

ianwallen committed:
Removed sequence name from annotation - issue will be resolved in version 4 of geonetwork (see https://github.com/geonetwork/core-geonetwork/pull/5003)
update migration to support sequence or hybernate_sequence

ianwallen commented:
@josegar74  as discussed I have removed the sequence name settings.  If we decide to merge PR #4781  then we can update that PR to include the sequence name.  Otherwise it should be fixed in GN 4 (#5003)

ianwallen committed:
Refactor - cleanup duplicate code for showing before/after versions.

ianwallen committed:
Fix encoding issue. - File should be UTF-8

ianwallen committed:
Titles will be null on migration so allow for null values. Otherwise jpa calls may fail.

ianwallen commented:
@josegar74 all the issues that you raised have been corrected.
Now that 3.10.4 has been released, I merged the latest master changes and then attempted to retest the migration logic to ensure that h2 migration was still working however I get an error 
```
org.h2.jdbc.JdbcSQLException: Unique index or primary key violation: "PRIMARY_KEY_842 ON PUBLIC.SETTINGS(NAME) VALUES ( /* 138 */ 'system/server/timeZone' )"; SQL statement:
 INSERT INTO Settings (name, value, datatype, position, internal) VALUES ('system/server/timeZone', '', 0, 260, 'n') [23505-174]

```
I noticed that  v3103 and v3104 are missing from the master.  So it seems that there was some steps missed when doing the releases.   I will need to wait for this to be fixed before I can continue with the testing.  Do you know who did the release? We should probably ask them to correct the issue so that they are aware of the missing step and hopefully not forget next time.

jodygarnett commented:
@ianwallen perhaps I am confused, the most recent release was from the `3.10.x` branch (and not `master`). What are `v3103` and `v3104` ?

ianwallen commented:
@jodygarnett 
We have upgraded our latest release to 3.10.4 tag.
This tag includes the v3103 and v3104 which are the migration script to upgrade to 3.10.3 and 3.10.4 as seen here

https://github.com/geonetwork/core-geonetwork/tree/3.10.4/web/src/main/webapp/WEB-INF/classes/setup/sql/migrate

Part of this PR is to change the keys so we are doing some of these changes as a migration process. Hence my test involve upgrading from a 3.10.4 release to master.

The problem is that the v3103 and v3104 does not exist and the changes that exists in v3103 and v3104 are in  v3110.
i.e. The following exist in both
`INSERT INTO Settings (name, value, datatype, position, internal) VALUES ('system/server/timeZone', '', 0, 260, 'n');`

So 3.10.4 already has the changes but v3110 is trying to apply the same changes again and causing failure.

I believe what is supposed to happen is the following.
During the release a there will be a release folder created - i.e. v3104
After the release 
v3104 is supposed to be added to master and any changes that exists in v3104 are supposed to be removed from v3110.

I'm quite sure @josegar74 is well aware of this and knows how to fix this.

ianwallen committed:
Added some corrections so that the migration works with PostgreSQL

ianwallen committed:
Added xsl to extra titles for iso19115-3.2018

ianwallen commented:
I have pulled and applied latest changes from the master. Applied changes from PR #5085 locally  and re-tested the migration for H2 and Postgres and they both seem to work as expected.

We should test on other databases if possible.

Note: Migration will only work if PR #5085 is merged in master.  Once the PR has been approved, I pull latest master changes into this PR for final testing.

jodygarnett commented:
Glad this is coming together, @ianwallen since releases are being discussed we should work with @fxprunayre and @josegar74 to schedule a `3.12.0` release for your team.

josegar74 commented on services/src/main/java/org/fao/geonet/api/records/MetadataWorkflowApi.java:
```diff
@@ -497,41 +687,219 @@ public MetadataStatus convertParameter(int id, MetadataStatusParameter parameter
+        UserSession user = ApiUtils.getUserSession(httpSession);
+        if (user.getProfile() != Profile.Administrator) {
+            if (groupId != null) {
+                final Specifications<UserGroup> spec = where(UserGroupSpecs.hasProfile(Profile.Editor))
```
To check if this code can be changed by `AccessManager.getGroups`

josegar74 commented on services/src/main/java/org/fao/geonet/api/records/MetadataWorkflowApi.java:
```diff
@@ -497,41 +687,219 @@ public MetadataStatus convertParameter(int id, MetadataStatusParameter parameter
+        UserSession user = ApiUtils.getUserSession(httpSession);
+        if (user.getProfile() != Profile.Administrator) {
+            if (groupId != null) {
+                final Specifications<UserGroup> spec = where(UserGroupSpecs.hasProfile(Profile.Editor))
```
To check if this code can be changed by `AccessManager.getGroups`

josegar74 commented on web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v3110/UpdateMetadataStatus.java:
```diff
@@ -0,0 +1,421 @@
+                        }
+                        if (metadataStatus.getTitles() == null || metadataStatus.getTitles().size() == 0) {
+                            LinkedHashMap<String, String> titles = titlesMap.get(metadataStatus.getMetadataId());
+                            // Try to get the titles from the schema.
```
Remove commented code

ianwallen committed:
Update permission check to use AccessManager.getGroups
Removed comments

josegar74 reviewed (approved):
Thanks for the updates, tested and looks working fine.

jodygarnett commented:
Notes from chat:
- Jose focused on `3.10.4.1` bug fix
- Will merge next week
- @ianwallen offered to cherry pick to `4.0.x` branch for @fxprunayre

ianwallen committed:
Merge branch 'master' of https://github.com/geonetwork/core-geonetwork into restorehistory_pk_fix

# Conflicts:
#	web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v3110/migrate-default.sql
#	web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v3110/migrate-mysql.sql
#	web/src/main/webapp/WEB-INF/classes/setup/sql/migrate/v3110/migrate-oracle.sql
